assignment-client/src/assets/UploadAssetTask.cpp

```
@@ -0,0 +1,80 @@
+//
+
+#include "UploadAssetTask.h"
+
+#include <QtCore/QBuffer>
+#include <QtCore/QFile>
+
+#include <AssetUtils.h>
+#include <NodeList.h>
+#include <NLPacketList.h>
+
+
+UploadAssetTask::UploadAssetTask(QSharedPointer<NLPacketList> packetList, SharedNodePointer senderNode,
+                                 const QDir& resourcesDir) :
+    _packetList(packetList),
+    _senderNode(senderNode),
+    _resourcesDir(resourcesDir)
+{
+    
+}
+
+void UploadAssetTask::run() {
+    auto data = _packetList->getMessage();
+    
+    QBuffer buffer { &data };
+    buffer.open(QIODevice::ReadOnly);
+    
+    MessageID messageID;
+    buffer.read(reinterpret_cast<char*>(&messageID), sizeof(messageID));
+    
+    uint8_t extensionLength;
+    buffer.read(reinterpret_cast<char*>(&extensionLength), sizeof(extensionLength));
+    
+    QByteArray extension = buffer.read(extensionLength);
+    
+    uint64_t fileSize;
+    buffer.read(reinterpret_cast<char*>(&fileSize), sizeof(fileSize));
+    
+    qDebug() << "UploadAssetTask reading a file of " << fileSize << "bytes and extension" << extension << "from"
+        << uuidStringWithoutCurlyBraces(_senderNode->getUUID());
+    
+    auto replyPacket = NLPacket::create(PacketType::AssetUploadReply);
+    replyPacket->writePrimitive(messageID);
+    
+    if (fileSize > MAX_UPLOAD_SIZE) {
+        replyPacket->writePrimitive(AssetServerError::AssetTooLarge);
+    } else {
+        QByteArray fileData = buffer.read(fileSize);
+        
+        auto hash = hashData(fileData);
+        auto hexHash = hash.toHex();
+        
+        qDebug() << "Hash for uploaded file from" << uuidStringWithoutCurlyBraces(_senderNode->getUUID())
+            << "is: (" << hexHash << ") ";
+        
+        QFile file { _resourcesDir.filePath(QString(hexHash)) + "." + QString(extension) };
+        
+        if (file.exists()) {
+            qDebug() << "[WARNING] This file already exists: " << hexHash;
```

shouldn't we return an error in this case?

```
 replyPacket->writePrimitive(AssetServerError::AssetAlreadyExists);
```